historical finds/Unravel/unravel_r2.2+/unravel.1
.\" %W%  %G%;
.TH UNRAVEL 1 "11 August 1995"
.SH NAME
unravel \- Compute program slices
.SH SYNOPSIS
.B unravel
.RB "[ "
.I directory
.RB " ]"
.SH DESCRIPTION

.LP
\fBUnravel\fR is a software analysis tool for extracting source program
statements relevant to a computation.  The computation is defined as
the state of some program variable at some statement in the program.
.LP
This is the main control window for \fBunravel\fR.
Before a slice can be
computed, the source files must be examined (analyzed) to build the
data structures needed to compute a program slice.
.LP
The main panel has several buttons to invoke each part of \fBunravel\fR,
review results and ask to see this help file.  The main panel also
displays the analysis status of each source file in the current
directory.
.LP
A typical \fBunravel\fR session would require the following:
.IP (1)
Preparation -- Locate the a directory containing the source files
to analyze.
.IP (2)
Run the \fBunravel\fR program -- This brings up the 
.I main control panel.
.IP (3)
Run the Analyzer -- Push the
.B Run Analyzer
button.  This brings up the
.I Analyzer Control Panel
where the files are selected for analysis.  From the
.I Analyzer Control Panel,
the button
.B Analyze Selected Files
will run the
analyzer on the selected files.  This builds data structures and
support files needed to perform slicing.  If any source file does not
contain ANSI C, \fBunravel\fR cannot analyze that file.  
.IP (4)
Review analysis results -- Examine the analysis results by pushing
the
.B Review History
button and selecting
.B Last Analysis
from the menu.
This brings up the results of the last run of the analyzer.
.IP (5)
Run the Slicer  -- Push the
.B Run Slicer
button.  This brings up the
.I Slice Control Panel.
From this panel you can calculate slices on the source
programs.
.SH FILES
.PD 0i
.IP \fIfile.\fBc\fR 1.5i
Source program
.IP \fIfile.\fBh\fR 1.5i
Source program header
.IP \fIfile.\fBLIF\fR 1.5i
Language Independent Format of \fIfile.\fBc\fR
.IP \fIfile.\fBT\fR 1.5i
Summary of \fIfile.\fBLIF\fR
.IP \fIfile.\fBH\fR 1.5i
Mapping of global variable declarations to header files
.IP \fBSYSTEM\fR 1.5i
Mapping of files to main programs
.IP \fIfile.\fBLINK\fR 1.5i
Link (merge) of \fBLIF\fR for all source files needed by main program
in \fIfile.\fBc\fR
.IP \fIfile.\fBK\fR 1.5i
Summary of \fIfile.\fBLINK\fR
.IP \fIfile.\fBY\fR 1.5i
Saved program slices from \fIfile.\fBc\fR
.IP \fBHISTORY\fR 1.5i
History of the current \fBunravel\fR session.
.IP \fBHISTOR-A\fR 1.5i
Results of last analyzer session
.IP \fBHISTORY-S\fR 1.5i
Results of last slice session
.IP \fBHISTORY.LOG\fR 1.5i
History of \fBunravel\fR sessions in the current directory.
.SH MORE INFORMATION
.PD .4v
\fBUnravel\fR was developed at:
.LP
Computer Systems Laboratory
.LP
National Institute of Standards and Technology
.br
Gaithersburg, Maryland 20899
.sp
The source code and a
two volumne technical report (NISTIR 5691) describing
the requirements
and design (volume 1) of \fBUnravel\fR along with a user manual
(volume 2) is available over the WWW at the following URL:
.br
http://hissa.ncsl.nist.gov/unravel.html
.br
For more information contact Dr. James R. Lyle (301) 975-3270


historical finds/Unravel/unravel_r2.2+/unravel.c
/*
*******************************************************************
*                                                                 *
*   Unravel control panel                                         *
*                                                                 *
*   Jim Lyle  9 June 1994                                         *
*                                                                 *
*******************************************************************
*/


# include <stdio.h>
# include <errno.h>
# include <X11/Intrinsic.h>
# include <X11/StringDefs.h>
# include <X11/Shell.h>
# include <X11/Xaw/Form.h>
# include <X11/Xaw/Command.h>
# include <X11/Xaw/Label.h>
# include <X11/Xaw/SimpleMenu.h>
# include <X11/Xaw/SmeBSB.h>
# include <X11/Xaw/SmeLine.h>
# include <X11/Xaw/MenuButton.h>
# include <X11/Xaw/AsciiText.h>
# include <sys/types.h>
# include <dirent.h>
# include <sys/stat.h>
# include <sys/time.h>
# define WAIT 500
# include "config.h"
# define XtNlinkOnly "linkOnly"

	static  char    sccsid[] = "@(#)unravel.c	1.8  7/26/96";

static int link_only = 0;
typedef struct {
	Widget	top,src,ana,not,out,prog,dup,link;
	Widget	text;
	char	*dir_buff;
} label_rec,*label_ptr;


Widget	help_label;

button_help (w,mess,e,ok)
	Widget	w;
	char	*mess;
	XEvent	*e;
	Boolean	*ok;
{
	XtVaSetValues (help_label,
		XtNlabel, (XtArgVal) mess,
		NULL);
}

set_button_help (w,mess)
	Widget	w;
	char	*mess;
{
	static char buff[] =
		"Description of object under mouse pointer is displayed here";
	XtAddEventHandler (w,EnterWindowMask,False,button_help,mess);
	XtAddEventHandler (w,LeaveWindowMask,False,button_help,buff);
}

int compare(a,b)
	struct dirent **a,**b;
{
	if (strcmp((*a)->d_name,(*b)->d_name) > 0) return 1;
	return 0;
}
int myscandir (dir_name,list,select,compare)
	char	*dir_name;
	struct dirent ***list;
	int (*select)();
	int (*compare)();
{
	DIR		*dir;
	struct dirent *entry;
	typedef struct l_struct ll_node,*ll;
	struct l_struct {
		ll		next;
		struct dirent *e;
	};

	int		i,n_match = 0;
	ll		new_ll;
	struct dirent *new_de;
	ll		nodes,node_list = NULL;

	dir = opendir (dir_name);
	if (!dir){
		fprintf (stderr,"directory open failed for %s\n",
			dir_name);
		exit(1);
	}
	for (entry = readdir(dir) ; entry != NULL ; entry = readdir(dir)){
		if (select(entry)){
			new_ll = (ll) malloc(sizeof (ll_node));
			new_ll->next = node_list;
			node_list = new_ll;
			new_ll->e = (struct dirent *) malloc (
				sizeof(struct dirent)+strlen(entry->d_name)+3);
			strcpy (new_ll->e->d_name,entry->d_name);
			n_match++;
		}
	}
	*list = (struct dirent **) malloc (
		(n_match + 1)*sizeof(struct dirent *));
	nodes = node_list;
	for (i = 0; i < n_match; i++){
		(*list)[i] = nodes->e;
		nodes = nodes -> next;
	}
	qsort (*list,n_match,sizeof(struct dirent *),compare);
	return n_match;
}

int file_selector (entry)
	struct dirent *entry;
{
	int		ix;

	ix = strlen (entry->d_name) - 2;

	if (ix > 0) 
		if ((entry->d_name[ix] == '.') &&
			(entry->d_name[ix+1] == 'c')){
				return 1;
		}
	return 0;
}

int scan_dir(f)
	struct dirent ***f;
{
	struct dirent **files;
	int		count;
	int		i;

	count = myscandir (".",&files,file_selector,compare);
	files[count] = NULL;
	*f = files;
	return count;
}

count_lif(n,f,n_ana,n_not,n_out)
	int		n,*n_ana,*n_not,*n_out;
	char	**f;
{
	int		nk = 0,k = 0,out = 0;
	struct	stat status_c,status_lif,status_t;
	int		c,lif,t;
	char	buff[2000];

	int		s,i;

	for (i = 0; i < n; i++){
		c = stat (f[i],&status_c);
		s = strlen(f[i]);
		strcpy (buff,f[i]);
		buff[s-1] = '\0';
		strcat (buff,"LIF");
		lif = stat (buff,&status_lif);
		strcpy (buff,f[i]);
		buff[s-1] = 'T';
		t = stat (buff,&status_t);
		if ((lif == 0) && (t == 0)){
			if ((status_c.st_mtime > status_t.st_mtime) ||
				(status_c.st_mtime > status_lif.st_mtime)) out++;
			else k++;
		}
		else nk++;

	}

	*n_ana = k;
	*n_not = nk;
	*n_out = out;
}

count_system(n_prog,n_dup,n_link)
	int		*n_prog,*n_dup,*n_link;
{
	int		n = 0,k,j,s,k_ok,link_ok;
	int		dup = 0;
	FILE	*sys;
	char	buff[2000],name[2000];
	int		ambig_state = 0;
	struct	stat status_k,status_link;

	n = 0;
	*n_link = 0;
	sys = fopen ("SYSTEM","r");
	if (sys){
		while (fgets (buff,2000,sys)){
			if (buff[0] == 'M'){
				ambig_state = 0;
				k = sscanf(buff,"MAIN %s %d",name,&k);
				if (k == 2){
					n++;
					strcpy (buff,name);
					s = strlen(name);
					buff[s-1] = '\0';
					strcat (buff,"LINK");
					link_ok = stat (buff,&status_link);
					buff[s-1] = '\0';
					strcat (buff,"K");
					k_ok = stat (buff,&status_k);
					if ((k_ok == 0) && (link_ok == 0))
						(*n_link)++;
				}
			}
			else if (buff[0] == 'A'){
				ambig_state = 1;
			}
			else if (buff[0] == 'F'){
				ambig_state = 0;
			}
			else if (buff[0] == 'L'){
				ambig_state = 0;
			}
			else if (ambig_state) dup++;
		}
		fclose(sys);
	}
		

	*n_prog = n;
	*n_dup = dup;
}

update (w,n,mess)
	Widget	w;
	int		n;
	char	*mess;
{
	char			buff[2000];

	sprintf (buff,mess,n);
	XtVaSetValues (w,XtNlabel,buff,NULL);
}

update_src_info(label)
	label_ptr	label;
{
	char			**file_names = NULL;
	struct dirent 	**files;
	int				n_files,i,n_analyzed,n_prog;
	int				n_not,n_out,n_dup,n_link;
	Widget	src,ana,not,out,prog,dup,link;

	src = label -> src;
	ana = label -> ana;
	not = label -> not;
	out = label -> out;
	prog = label -> prog;
	dup = label -> dup;
	link = label -> link;
	n_files = scan_dir(&files);
	file_names = (char **) malloc ((n_files + 1)*sizeof (char *));
	for (i = 0; files[i] ; i++){
		file_names[i] = files[i]->d_name;
	}
	file_names[n_files] = NULL;
	count_lif(n_files,file_names,&n_analyzed,&n_not,&n_out);
	count_system(&n_prog,&n_dup,&n_link);

	update (ana,n_analyzed,n_analyzed == 1?"%3d file analyzed & up to date":
			"%3d files analyzed & up to date");
	update (src,n_files,n_files == 1?"%3d source file":
			"%3d source files");
	update (not,n_not,n_not == 1?"%3d source file not analyzed":
			"%3d source files not analyzed");
	update (out,n_out,n_out == 1?"%3d file analyzed & out of date":
			"%3d files analyzed & out of date");
	update (link,n_link,n_link == 1?"%3d main program file linked":
			"%3d main program files linked");
	update (prog,n_prog,n_prog == 1?"%3d main program file analyzed":
			"%3d main program files analyzed");
	update (dup,n_dup,n_dup == 1?"%3d duplicate procedure found":
			"%3d duplicate procedures found");
}


ana_done  (w,labels,e,ok)
	Widget	w;
	XEvent	*e;
	label_ptr	labels;
	Boolean	*ok;
{
	update_src_info(labels);
}

push_analysis(w,labels,w_data)
	Widget	w;
	label_ptr	labels;
	XtPointer	w_data;
{
	char	buff[2000];
	static int		set = 0;

	if (!set) {
		set = 1;
		XtAddEventHandler (w, NoEventMask,True,ana_done,labels);
	}
	sprintf (buff,"PATH=%s:$PATH ; analyzer %d &",HOME,XtWindow(w));
	system (buff);
	update_src_info(labels);
}

push_review(w,hist,w_data)
	Widget	w;
	int			hist;
	XtPointer	w_data;
{
	char	buff[2000];
	static 	char *title[] = {
				"Result of Last Analysis",
				"Last Slices Requested",
				"Log of Current Unravel Session",
				"Unravel Log of this Directory"
			},
			*file_name[] = {
				"HISTORY-A",
				"HISTORY-S",
				"HISTORY",
				"HISTORY.LOG"
			};
	sprintf (buff,"%s/helpu %s -geom +10+10 -title \"%s\" %s &",
		HOME,"-xrm \"*Command.label: History Review Done\"",
		title[hist],file_name[hist]);
	system(buff);
}

push_slice(w,u_data,w_data)
	Widget	w;
	XtPointer	u_data;
	XtPointer	w_data;
{
	char	buff[2000];
	sprintf (buff,"%s/select %s >HISTORY-S 2>&1 &",HOME,
		link_only?"-noslice":"");
	system (buff);
}

push_dir(w,labels,w_data)
	Widget	w;
	XtPointer	w_data;
	label_ptr	labels;
{
	char	buff[2000];
	int		status;
	XawTextPosition	from,to;
	XawTextBlock	block;


	status = chdir (labels->dir_buff);
	if (status){
		strcpy (buff,(char *)getcwd(NULL,1000));
		block.firstPos = 0;
		block.length = strlen (buff);
		block.ptr = buff;
		block.format = FMT8BIT;

		from = 0;
		to = strlen (labels->dir_buff);
		/*
		XawTextReplace (labels->text,from,to,&block);
		XawTextSetInsertionPoint (labels->text,block.length);
		*/
	}
	else {
		update_src_info(labels);
		sprintf (buff,
			"echo -n \"Unravel change directory: %s \" >> HISTORY",
			labels->dir_buff);
		system(buff);
		system("date >> HISTORY");

	}
}

push_help(w,u_data,w_data)
	Widget	w;
	XtPointer	u_data;
	XtPointer	w_data;
{
	char	buff[2000];

	sprintf (buff,"%s/helpu -geom +10+10 %s/unravel.help %s &",HOME,HOME,
		"-title \"Unravel Main Help\" ");
	system (buff);
}

done(w,u_data,w_data)
	Widget	w;
	XtPointer	u_data;
	XtPointer	w_data;
{
	system ("/bin/cat HISTORY >> HISTORY.LOG; /bin/rm -f HISTORY *ERR");
	exit (0);
}

strech (goal,w)
	Widget	goal,w;
{
	Dimension		width,bw;
	Position		x;
	int				dist;

	XtUnmanageChild(w);
	XtVaGetValues (goal,
		XtNdefaultDistance, &dist,
		XtNwidth,&width,NULL);
	XtVaGetValues (w,
		XtNborderWidth,&bw,
		XtNx,&x,
		NULL);
	XtVaSetValues (w,
		XtNwidth,	(XtArgVal) (width-2*bw - dist - x),
		NULL);
	XtManageChild(w);
}

make_windows(top)
	Widget	top;
{
	Widget			analysis,review,slice,help,quit;
	Widget			info,dir,dir_label;
	Widget			src,ana,not,out,prog,dup;
	Widget			frame,link;
	Widget			review_menu,review_ana,review_slice,
					review_all,review_this;
	static 	label_rec labels;
	static 	char	dir_buff[2000] = "";
	char			buff[2000];

	labels.top = top;
	labels.dir_buff = dir_buff;
	frame = XtVaCreateManagedWidget ("frame",formWidgetClass,
		top,NULL);

	quit = XtVaCreateManagedWidget ("quit",commandWidgetClass,
		frame,
		/*
		XtNlabel,			"Exit Unravel",
		*/
		XtNtop,				XawChainTop,
		XtNbottom,			XawChainTop,
		XtNleft,			XawChainLeft,
		XtNright,			XawChainLeft,
		NULL);
	XtAddCallback(quit,XtNcallback,done,NULL);

	analysis = XtVaCreateManagedWidget ("analysis",commandWidgetClass,
		frame,
		XtNtop,				XawChainTop,
		XtNbottom,			XawChainTop,
		XtNleft,			XawChainLeft,
		XtNright,			XawChainLeft,
		XtNfromHoriz,		quit,
		NULL);
	XtAddCallback(analysis,XtNcallback,push_analysis,&labels);

	review_menu = XtVaCreatePopupShell ("reviewMenu",
		simpleMenuWidgetClass,top,NULL);
	
	review = XtVaCreateManagedWidget ("review",
		menuButtonWidgetClass,frame,
		XtNlabel,			"Review History",
		XtNtop,				XawChainTop,
		XtNbottom,			XawChainTop,
		XtNleft,			XawChainLeft,
		XtNright,			XawChainLeft,
		XtNfromHoriz,		analysis,
		XtNmenuName,	(XtArgVal) "reviewMenu",
		NULL);
	review_ana = XtVaCreateManagedWidget ("reviewAnalysis",
		smeBSBObjectClass,review_menu,
		XtNlabel,		(XtArgVal) "Last Analysis",
		NULL);
	review_slice = XtVaCreateManagedWidget ("reviewSlice",
		smeBSBObjectClass,review_menu,
		XtNlabel,		(XtArgVal) "Last Slice",
		NULL);
	review_this = XtVaCreateManagedWidget ("review",
		smeBSBObjectClass,review_menu,
		XtNlabel,		(XtArgVal) "This Session",
		NULL);
	review_all = XtVaCreateManagedWidget ("reviewAll",
		smeBSBObjectClass,review_menu,
		XtNlabel,		(XtArgVal) "All History",
		NULL);
	XtAddCallback(review_ana,XtNcallback,push_review,0);
	XtAddCallback(review_slice,XtNcallback,push_review,1);
	XtAddCallback(review_this,XtNcallback,push_review,2);
	XtAddCallback(review_all,XtNcallback,push_review,3);

	slice = XtVaCreateManagedWidget ("slice",commandWidgetClass,
		frame,
		XtNtop,				XawChainTop,
		XtNbottom,			XawChainTop,
		XtNleft,			XawChainLeft,
		XtNright,			XawChainLeft,
		XtNfromHoriz,		review,
		NULL);
	XtAddCallback(slice,XtNcallback,push_slice,NULL);

	help = XtVaCreateManagedWidget ("help",commandWidgetClass,
		frame,
		XtNlabel,			"Help",
		XtNtop,				XawChainTop,
		XtNbottom,			XawChainTop,
		XtNleft,			XawChainRight,
		XtNright,			XawChainRight,
		XtNfromHoriz,		slice,
		NULL);
	XtAddCallback(help,XtNcallback,push_help,NULL);

	strcpy (buff,"Current directory: ");

	strcat (buff,getcwd(NULL,1000));
	dir = XtVaCreateManagedWidget ("dir",labelWidgetClass,
		frame,
		XtNlabel,			buff,
		XtNtop,				XawChainTop,
		XtNbottom,			XawChainTop,
		XtNleft,			XawChainLeft,
		XtNright,			XawChainRight,
		XtNfromVert,		quit,
		XtNjustify,			XtJustifyLeft,
		NULL);
	/*
	dir_label = XtVaCreateManagedWidget ("dirLabel",
		commandWidgetClass,
		frame,
		XtNlabel,			buff,
		XtNtop,				XawChainTop,
		XtNbottom,			XawChainTop,
		XtNleft,			XawChainLeft,
		XtNright,			XawChainLeft,
		XtNfromVert,		quit,
		XtNjustify,			XtJustifyLeft,
		NULL);
	
	XtAddCallback(dir_label,XtNcallback,push_dir,&labels);
	strcpy (dir_buff,getcwd(NULL,1000));
	dir = XtVaCreateManagedWidget ("dir",
		asciiTextWidgetClass,
		frame,
		XtNfromHoriz,		(XtArgVal) dir_label,
		XtNfromVert,		(XtArgVal) quit,
		XtNleft,			(XtArgVal) XawChainLeft,
		XtNright,			(XtArgVal) XawChainRight,
		XtNtop,				(XtArgVal) XawChainTop,
		XtNbottom,			(XtArgVal) XawChainTop,
		XtNborderWidth,		(XtArgVal) 0,
		XtNautoFill,		(XtArgVal) False,
		XtNdisplayNonprinting, (XtArgVal) True,
		XtNeditType,		(XtArgVal) XawtextEdit,
		XtNlength,			(XtArgVal) 1000,
		XtNstring,			(XtArgVal) dir_buff,
		XtNuseStringInPlace,(XtArgVal) True,
		NULL);
	XawTextSetInsertionPoint (dir,strlen(dir_buff));
	*/
	labels.text = dir;

	src = XtVaCreateManagedWidget ("src",labelWidgetClass,
		frame,
		XtNtop,				XawChainTop,
		XtNbottom,			XawChainTop,
		XtNleft,			XawChainLeft,
		XtNright,			XawChainRight,
		XtNfromVert,		dir,
		XtNjustify,			XtJustifyLeft,
		NULL);

	ana = XtVaCreateManagedWidget ("ana",labelWidgetClass,
		frame,
		XtNtop,				XawChainTop,
		XtNbottom,			XawChainTop,
		XtNleft,			XawChainLeft,
		XtNright,			XawChainRight,
		XtNfromVert,		src,
		XtNjustify,			XtJustifyLeft,
		NULL);

	not = XtVaCreateManagedWidget ("not",labelWidgetClass,
		frame,
		XtNtop,				XawChainTop,
		XtNbottom,			XawChainTop,
		XtNleft,			XawChainLeft,
		XtNright,			XawChainRight,
		XtNfromVert,		ana,
		XtNjustify,			XtJustifyLeft,
		NULL);

	out = XtVaCreateManagedWidget ("out",labelWidgetClass,
		frame,
		XtNtop,				XawChainTop,
		XtNbottom,			XawChainTop,
		XtNleft,			XawChainLeft,
		XtNright,			XawChainRight,
		XtNfromVert,		not,
		XtNjustify,			XtJustifyLeft,
		NULL);

	prog = XtVaCreateManagedWidget ("prog",labelWidgetClass,
		frame,
		XtNtop,				XawChainTop,
		XtNbottom,			XawChainTop,
		XtNleft,			XawChainLeft,
		XtNright,			XawChainRight,
		XtNfromVert,		out,
		XtNjustify,			XtJustifyLeft,
		NULL);

	link = XtVaCreateManagedWidget ("link",labelWidgetClass,
		frame,
		XtNtop,				XawChainTop,
		XtNbottom,			XawChainTop,
		XtNleft,			XawChainLeft,
		XtNright,			XawChainRight,
		XtNfromVert,		prog,
		XtNjustify,			XtJustifyLeft,
		NULL);

	dup = XtVaCreateManagedWidget ("dup",labelWidgetClass,
		frame,
		XtNtop,				XawChainTop,
		XtNbottom,			XawChainTop,
		XtNleft,			XawChainLeft,
		XtNright,			XawChainRight,
		XtNfromVert,		link,
		XtNjustify,			XtJustifyLeft,
		NULL);

	info = XtVaCreateManagedWidget ("info",labelWidgetClass,
		frame,
		XtNlabel,
		"Description of object under mouse pointer is displayed here",
		XtNtop,				XawChainTop,
		XtNbottom,			XawChainTop,
		XtNleft,			XawChainLeft,
		XtNright,			XawChainRight,
		XtNfromVert,		dup,
		XtNjustify,			XtJustifyLeft,
		NULL);

	labels.src = src;
	labels.ana = ana;
	labels.not = not;
	labels.out = out;
	labels.prog = prog;
	labels.dup = dup;
	labels.link = link;
	update_src_info(&labels);

	XtRealizeWidget(top);
	strech (frame,ana);
	strech (frame,dir);
	strech (frame,info);
	strech (frame,not);
	strech (frame,out);
	strech (frame,prog);
	strech (frame,link);
	strech (frame,dup);
	strech (frame,src);

	help_label = info;
	set_button_help (quit,"Click to exit when finished");
	set_button_help (help,
		"Click to pop up a window describing unravel");
	set_button_help (ana,
		"Display how many source files have been analyzed");
	set_button_help (not,
		"Display how many source files have not been analyzed");
	set_button_help (out,
		"Display how many source files have been modified since being analyzed");
	set_button_help (slice,"Click to setup and run the program slicer");
	set_button_help (review,"Click to pop up menu of history displays");
	set_button_help (analysis,"Click to setup and run the analyzer");
	set_button_help (link, "Display # of main programs linked");
	set_button_help (prog, "Display # of files analyzed \
		containing a main procedure");
	set_button_help (dup,"Display how many procedures are unresolved");
	set_button_help (src,
		"Display how many source files are in the directory");
	set_button_help (dir,
		"Display the current directory name");
	/*
	set_button_help (dir,
		"Display the current directory name or edit new name");

	set_button_help (dir_label,"Click to change directory");
	*/
	set_button_help (frame,"   . . .");
	XtInstallAccelerators (frame,quit);
	XtInstallAccelerators (frame,help);
	/*
	XtInstallAccelerators (dir,dir_label);
	*/

}

char *fall[] = {
		/* don't move the first entry (iconPixmap) */
		"*iconPixmap:	unravel.icon",
		"*font:	lucidasans-bold-14",
		"*background: seashell3",
		"*foreground: blue",
		"*shapeStyle: oval",
		"*Label*borderWidth: 0",
		"*Command*borderWidth: 2",
		"*MenuButton*borderWidth: 2",
		"*info*borderWidth: 3",
		"*quit.label: Exit Unravel",
		"*analysis.label: Run Analyzer",
		"*slice.label: Run Slicer",
		"*title: Unravel Version 2.2+ Main Control Panel",
		"*geometry: +10+10",
		"*quit.accelerators: #override\\n \
			<KeyPress>q: set() highlight() notify()",
		"*help.accelerators:  \
			<KeyPress>h: set() highlight() notify() unset()\\n\
			<KeyPress>?: set() highlight() notify() unset()",
		/*
		"*dirLabel.accelerators: #override\\n\
			<KeyPress>Return: set() highlight() notify() unset()\\n\
			<LeaveNotify>: set() highlight() notify() unset()",
		*/
	NULL};

/*
# define XtNrunningFG "runningFG"
# define XtNrunningBG "runningBG"
	static XrmOptionDescRec opt[] = {
		{"-runningfg", "*runningFG", XrmoptionSepArg,NULL},
		{"-runningbg", "*runningBG", XrmoptionSepArg,NULL}};
	static XtResource res[] = {
		{XtNrunningFG,XtCForeground, XtRPixel, sizeof(Pixel),
			XtOffsetOf(struct  app_resources,fg),XtRString,"red"},
		{XtNrunningBG,XtCForeground, XtRPixel, sizeof(Pixel),
			XtOffsetOf(struct  app_resources,bg),XtRString,"yellow"}};
	top = XtAppInitialize (&ac,"Xtform",opt,XtNumber(opt),&np,p,fall,NULL,0);
	XtGetApplicationResources (top,&a_res,res,XtNumber(res),NULL,0);
	
*/
	static XtResource res[] = {
		{XtNlinkOnly,XtCValue, XtRInt, sizeof(int),
			0,XtRString,"0"}};

	static XrmOptionDescRec opt[] = {
		{"-noslice", "*linkOnly", XrmoptionNoArg,"1"}};
main (np,p)
	int		np;
	char	*p[];
{
	XtAppContext	ac;
	Widget			top;
	char		**file_names;
	struct dirent 	**files;
	int				n_files,i,n_analyzed,n_prog;
	int				n_not,n_out;
	char			icon_at[2000],cmd[2000];
	int				sys_ok;

	fprintf (stdout,"TEST OUTPUT 0 in unravel.c sgw\n");

	sprintf (icon_at,"*iconPixmap: %s/unravel.icon",HOME);
	fall[0] = icon_at;

	top = XtAppInitialize (&ac,"Unravel",opt,XtNumber(opt),&np,p,
		fall,NULL,0);

	fprintf (stdout,"TEST OUTPUT 1 in unravel.c sgw\n");

	XtGetApplicationResources (top,&link_only,res,XtNumber(res),NULL,0);
	if (np == 2){
		/* printf ("changing to %s\n",p[1]); */
		errno = 0;
		i = chdir (p[1]);
		/* printf ("chdir returned %d %d\n",i,errno); */
		if (i != 0){
			fprintf (stderr,"%s could not change directory to %s\n",
				p[0],p[1]);
			perror("");
			exit(1);
		}
	}

	fprintf (stdout,"TEST OUTPUT 2 in unravel.c sgw\n");

	sys_ok = system ("echo \" \" >HISTORY");
	if (sys_ok != 0){
		fprintf (stderr,"%s could not write in file directory\n",
				p[0]);
		fprintf (stderr,"continue? [y|n] ");
		scanf ("%c",cmd);
		if (cmd[0] != 'y') exit(1);
	}

	fprintf (stdout,"TEST OUTPUT 3 in unravel.c sgw\n");

	sprintf (cmd,"echo -n UNRAVEL \" %s \" >> HISTORY ; date >>HISTORY",
		getcwd (NULL,2000));

	fprintf (stdout,"TEST OUTPUT 4 in unravel.c sgw\n");

	system(cmd);

	fprintf (stdout,"TEST OUTPUT 5 in unravel.c sgw\n");

	system ("echo \"No slices computed this session\" >HISTORY-S");
	system ("echo \"No analysis done this session\" >HISTORY-A");

	fprintf (stdout,"TEST OUTPUT 6 in unravel.c sgw\n");

	make_windows(top);

	fprintf (stdout,"TEST OUTPUT 7 in unravel.c sgw\n");

	XtAppMainLoop(ac);

	fprintf (stdout,"TEST OUTPUT 8 in unravel.c sgw\n");
}
